cpp/src/table/row_operators.cu

```diff
@@ -77,50 +149,75 @@ auto decompose_structs(table_view table,
                        host_span<order const> column_order         = {},
                        host_span<null_order const> null_precedence = {})
 {
+  auto sliced         = pushdown_struct_offsets(table);
+  auto linked_columns = detail::table_to_linked_columns(sliced);
+
   std::vector<column_view> verticalized_columns;
   std::vector<order> new_column_order;
   std::vector<null_order> new_null_precedence;
   std::vector<int> verticalized_col_depths;
-  for (size_type col_idx = 0; col_idx < table.num_columns(); ++col_idx) {
-    auto const& col = table.column(col_idx);
-    if (is_nested(col.type())) {
+  for (size_t col_idx = 0; col_idx < linked_columns.size(); ++col_idx) {
+    auto const col = linked_columns[col_idx];
+    if (is_nested(col->type())) {
       // convert and insert
       std::vector<column_view> r_verticalized_columns;
       std::vector<int> r_verticalized_col_depths;
-      std::vector<column_view> flattened;
+      std::vector<detail::LinkedColPtr> flattened;
       std::vector<int> depths;
       // TODO: Here I added a bogus leaf column at the beginning to help in the while loop below.
       //       Refactor the while loop so that it can handle the last case.
-      flattened.push_back(make_empty_column(type_id::INT32)->view());
-      std::function<void(column_view const&, int)> recursive_child = [&](column_view const& c,
-                                                                         int depth) {
+      flattened.push_back(
+        std::make_shared<detail::linked_column_view>(make_empty_column(type_id::INT32)->view()));
```

Comment:
Why does `flattened` need to be an owner of the resulting column views, i.e. why does it need shared pointers rather than just references? I'm not following the need for shared ownership yet. If you created this "bogus leaf column" and stored it, then would it be possible for the contents of `flattened` to just be references into the contents of `linked_columns` so that `linked_columns` could then just be a `std::vector<linked_column_view>` instead of `std::vector<std::shared_ptr<linked_column_view>>`?

Reply:
They can be references. I just though shared pointers are cheap enough at this scale.

Reply:
Changed to storing raw pointers https://github.com/rapidsai/cudf/pull/10289/commits/046c407df04aed8f33419c75cda0844411263e7d